Java: Add ability to use different OpenAI models for semantic functions

relango opened the pull request:
### Motivation and Context

Link to [issue](https://github.com/microsoft/semantic-kernel/issues/2387). 

### Description

Here is how I added multi model support

- Added `serviceId` field to [PromptTemplateConfig](https://github.com/microsoft/semantic-kernel/blob/experimental-java/java/semantickernel-api/src/main/java/com/microsoft/semantickernel/semanticfunctions/PromptTemplateConfig.java) so that `serviceId` configured in `config.json` can be read.
- Fixed [DefaultCompletionSKFunction](https://github.com/microsoft/semantic-kernel/blob/experimental-java/java/semantickernel-core/src/main/java/com/microsoft/semantickernel/orchestration/DefaultCompletionSKFunction.java#L240) to use `serviceId` from promptTemplateConfig associated with SemanticFunctionConfig instead of hardcoded `null`

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

dsgrieve commented on java/semantickernel-api/src/main/java/com/microsoft/semantickernel/semanticfunctions/PromptTemplateConfig.java:
```diff
@@ -346,26 +346,34 @@ public List<InputParameter> getParameters() {
             @JsonProperty("schema") int schema,
             @JsonProperty("description") String description,
             @JsonProperty("type") String type,
+            @JsonProperty("serviceId") String serviceId,
```
The convention set by .NET is to use snake case. 

PromptTemplateConfig.cs has List<string> DefaultServices. Maybe @shawncal could weigh in on that.

relango replied:
Yeah, I could not find an equivalent c# or Python implementation. I can change based on the feedback. let me know

relango replied:
@shawncal @dsgrieve Looks like this is the only outstanding comment. Can we get a consensus on this so that this PR can be merged?

markwallace-microsoft replied:
@relango Apologies for the delay, few folks have been out. I'm taking a look at this now. We want to align this across all of the implementations.

markwallace-microsoft replied:
@relango Can you take a look at this PR https://github.com/microsoft/semantic-kernel/pull/2739

We want to put the `serviceId` in the completion config. The reason for this is we want to support multiple services within a semantic function and allow for the appropriate service to be selected when the prompt is executed. 

If you update your PR to follow the same pattern I can get it approved and merged.

relango replied:
Thank you @markwallace-microsoft , I looked at the PR [#2739 ](https://github.com/microsoft/semantic-kernel/pull/2739) , I already have `serviceId` added to `PromptTemplateConfig` in this PR. 

The only change I see is JsonProperty name should be changed to `service_id`. I will update PR with that change. Let me know if I missed anything.

relango replied:
@markwallace-microsoft I updated PR to change JsonProperty to `service_id` in `PromptTemplateConfig`. Let me know if any other changes are needed.

relango replied:
@markwallace-microsoft @johnoliver  I have rebased my PR with the latest commit. Can we get it reviewed and merged? Let me know if any additional changes are needed.

dsgrieve reviewed (commented):
I defer to @johnoliver

relango commented:
@microsoft-github-policy-service agree company="Salesforce"

relango commented on java/samples/sample-code/src/main/java/com/microsoft/semantickernel/samples/Example06_UseMultipleModels.java:
```diff
@@ -0,0 +1,83 @@
+ * started with Azure OpenAI</a>
+ * <a href="https://openai.com/product">Get started with OpenAI</a>
+ */
+public class Example06_UseMultipleModels {
```
Updated to `Example11_UseMultipleModels`

johnoliver commented:
I think the pattern we need to replicate is https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatCompletionServiceExtensions.cs#L22 where when they get a ChatCompletion, they provide an optional serviceId.

Since we use builders Java side, IMO the equivalent translation for this is to add setServiceId to CompletionSKFunction.Builder then pass that serviceId on to DefaultCompletionSKFunction and then select our service based on that id at: https://github.com/microsoft/semantic-kernel/blob/experimental-java/java/semantickernel-core/src/main/java/com/microsoft/semantickernel/orchestration/DefaultCompletionSKFunction.java#L238

relango commented:
@johnoliver When skills are imported to kernel, the `DefaultCompletionSKFunction`  builder is abstracted and configurations are loaded using `config.json` file. That's why I added `serviceId` to PromptTemplateConfig. 

Can you look at the [example code](https://github.com/microsoft/semantic-kernel/blob/9393af29d82ffdd659a71a1b94735a08463c8b55/java/samples/sample-code/src/main/java/com/microsoft/semantickernel/samples/Example11_UseMultipleModels.java) I added and let me know how we will able to pass serviceId if it is added to  `DefaultCompletionSKFunction`  builder ?

johnoliver commented:
@relango Ok I see the issue, yes my suggestion would not work without probably larger changes to other parts of the api.

@shawncal this is a change that would impact the other languages too, do you think this is the way to go, is there a different way around this? #1787 suggests adding preferred_service, and #989 suggests an alternative that I don't think would provide a solution to skills loaded externally.

markwallace-microsoft reviewed (commented):
Please refactor to match https://github.com/microsoft/semantic-kernel/pull/2739

relango commented:
Thank you for approval @markwallace-microsoft , can we get it merged?

relango commented:
@johnoliver @markwallace-microsoft 

Just following up on this PR. Is there anything need to get this PR merged?

johnoliver commented:
@markwallace-microsoft it seems that you added `service_id` in https://github.com/microsoft/semantic-kernel/pull/2739 but I cant find that code now, was it intentionally removed or has that code moved? 

@relango I will need to test, but looking at .NET side, it seems like they are adding service_id to the "completion" property, i.e looking at:

https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/IntegrationTests/Connectors/OpenAI/OpenAICompletionTests.cs#L348

```
            {
                "completion": {
                    "max_tokens": 256,
                    "service_id"": "azure-text-davinci-003"
                }
            }
```

Presumably this means that service_id us adding it to PromptTemplateConfig would be a break from .net schema. 


@markwallace-microsoft  can you confirm if .net has implemented
```
            {
                "completion": {
                    "max_tokens": 256,
                    "service_id"": "azure-text-davinci-003"
                }
            }
```

OR

```
            {
                "completion": {
                    "max_tokens": 256
                },
                "service_id"": "azure-text-davinci-003"
            }
```

relango commented:
Thank you for update @johnoliver. Please let me know if you need anything from me. Appreciate if this change could merged sooner.

brunoborges commented:
@johnoliver please follow up on this before the next release of SK for Java

johnoliver commented:
@relango I have opened a PR on your branch at:

https://github.com/relango/semantic-kernel/pull/1

This PR updates for full support for ADR 008 (https://github.com/microsoft/semantic-kernel/blob/main/docs/decisions/0008-support-generic-llm-request-settings.md), which is the ADR in this area.

This PR does the following:
- Bring your branch up to date with java-development branch (you will also need to change the branch this PR is against to the java-development branch)
- Updates the code to support ADR.
- Apply code formatting needed on these

If you dont want do do this PR via your branch I can PR it directly and close this PR. 


You can see a slightly cleaner view of the changes at:

https://github.com/microsoft/semantic-kernel/compare/java-development...johnoliver:semantic-kernel:experimental-java-sfdc-multimodel-support-4?expand=1&w=1

The key examples of the usage can be seen at: https://github.com/microsoft/semantic-kernel/compare/java-development...johnoliver:semantic-kernel:experimental-java-sfdc-multimodel-support-4?expand=1&w=1#diff-c26be3bc48c2d3fd4faf320f234e13d113f0464987c4e023723a619ae56b4de0R32

Now you can add serviceId and model to invocation, i.e:

```
Mono<SKContext> result = function.invokeAsync("time travel to dinosaur age",
        SKBuilders
            .completionRequestSettings()
            .serviceId("text-davinci-003")
            .build());
```

function config, i.e:


```

    var function = kernel.
        getSemanticFunctionBuilder()
        .withPromptTemplate("tell a joke about {{$input}}")
        .withRequestSettings(
            SKBuilders.completionRequestSettings()
                .temperature(0.4)
                .topP(1)
                .maxTokens(100)
                .serviceId("text-davinci-003")
                .build())
        .build();
```

Or config file

```
{
  "schema": 1,
  "description": "Sample plugin with a model and service id",
  "type": "completion",
  "completion": {
    "max_tokens": 150,
    "temperature": 0.9,
    "top_p": 0.0,
    "presence_penalty": 0.6,
    "frequency_penalty": 0.0,
    "service_id": "text-davinci-003",
    "model_id": "text-davinci-003"
  }
}
```

Let me know what you think of the changes and if they address your usecase. If you are ok with it you can merge the changes into your branch and I can get someone here to review it (as my code is in there will need a different reviewer)

relango commented:
Thank you @johnoliver . The code changes look good. Since you already got my changes in your branch/PR. Please go ahead merge yours and close this PR.

Let me know once it is merged, and I will test it.

johnoliver commented:
Closing in favour of #3366
